scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,324 @@
+  match_cmd = "org.apache.zookeeper.server.quorum.QuorumPeerMain"
+  monit_app_configuration.create_config_file(ZK_WATCH_NAME, start_cmd, stop_cmd,
+    ports=[zk.DEFAULT_PORT], upgrade_flag=True, match_cmd=match_cmd)
+
+  if not monit_interface.start(ZK_WATCH_NAME):
+    logger.error("Monit was unable to start ZooKeeper.")
+    sys.exit(1)
+  else:
+    logger.info("Successfully started ZooKeeper.")
+
+def validate_and_update_entities(datastore, ds_distributed):
+  """ Validates entities in batches of BATCH_SIZE, deletes tombstoned
+  entities (if any) and updates invalid entities.
+  Args:
+    datastore: A reference to the batch datastore interface.
+    ds_distributed: A reference to the distributed datastore.
+  """
+  last_key = ""
+  entities_checked = 0
+  last_logged = time.time()
+  while True:
+    try:
+      logger.debug("Fetching {} entities".format(BATCH_SIZE))
+      entities = get_entity_batch(last_key, datastore, BATCH_SIZE)
+
+      for entity in entities:
+        process_entity(entity, datastore, ds_distributed)
+
+      last_key = entities[-1].keys()[0]
+      entities_checked += len(entities)
+
+      if time.time() > last_logged + LOG_PROGRESS_FREQUENCY:
+        logger.info("Checked {} entities".format(entities_checked))
+        last_logged = time.time()
+
+      if len(entities) < BATCH_SIZE:
+        break
+
+    except datastore_errors.Error as error:
+      logger.error("Error getting a batch of entities: {}".format(error))
+    except AppScaleDBConnectionError as connection_error:
+      logger.error("Error getting a batch of entities: {}".format(connection_error))
+
+def get_entity_batch(last_key, datastore, batch_size):
+  """ Gets a batch of entities to operate on.
+  Args:
+    last_key: The last key from a previous query.
+    datastore: A reference to the batch datastore interface.
+    batch_size: The max number of entities retrieved in the datastore request.
+  Returns:
+    A list of entities with a limit of batch_size.
+  """
+  return datastore.range_query(APP_ENTITY_TABLE, APP_ENTITY_SCHEMA, last_key,
+    "", batch_size, start_inclusive=False)
+
+def process_entity(entity, datastore, ds_distributed):
+  """ Processes an entity by updating it if necessary and removing tombstones.
+  Args:
+    entity: The entity to process.
+    datastore: A reference to the batch datastore interface.
+    ds_distributed: A reference to the distributed datastore.
+  Returns:
+    True on success, False otherwise.
+  """
+  logger.debug("Process entity {}".format(str(entity)))
+  key = entity.keys()[0]
+  one_entity = entity[key][APP_ENTITY_SCHEMA[0]]
+  version = entity[key][APP_ENTITY_SCHEMA[1]]
+  if one_entity == datastore_server.TOMBSTONE:
+    return delete_entity_from_table(key, datastore)
+
+  entity_proto = entity_pb.EntityProto(one_entity)
+  app_id = entity_proto.key().app()
+  validated_entity = ds_distributed.validated_result(app_id, entity)
+
+  if not validated_entity:
+    return True
```

Comment:
If `validated_entity` is an empty dict, that means that there's no valid transaction for the entity. It should be removed from the entities table.